src/Board.cpp

```diff
@@ -0,0 +1,168 @@
+{
+    return !isEmpty(source) && !isRival(source);
+}
+
+bool Board::isWayBlocked(const Location& source, const Location& destination) const
+{
+    int dx = destination.x - source.x;
+    int dy = destination.y - source.y;
+
+    if (dx != 0 && dy != 0 && std::abs(dx) != std::abs(dy))
+        return false;
+
+    int x_step = (dx > 0) ? 1 : ((dx < 0) ? -1 : 0);
+    int y_step = (dy > 0) ? 1 : ((dy < 0) ? -1 : 0);
+    Location currentLoc = source;
+    currentLoc.x += x_step;
+    currentLoc.y += y_step;
+    while (currentLoc.x != destination.x || currentLoc.y != destination.y) {
+        if (!isEmpty(currentLoc)) {
+            return true;
+        }
+        currentLoc.x += x_step;
+        currentLoc.y += y_step;
+    }
+    return false;
+
+}
+
+bool Board::moveCausesChess(const Location& source, const Location& destination)
+{
+    auto temp = std::move(m_board[destination.x][destination.y]);
+    m_board[destination.x][destination.y] = std::move(m_board[source.x][source.y]);
+
+    if (isKing(destination))
+        m_kingLoc[m_board[destination.x][destination.y]->getColor()] = destination;
+        
+    bool isThreatened = mate();
+
+    if (isKing(destination))
+        m_kingLoc[m_board[destination.x][destination.y]->getColor()] = source;
+       
+    m_board[source.x][source.y] = std::move(m_board[destination.x][destination.y]);
+    m_board[destination.x][destination.y] = std::move(temp);
+    return isThreatened;
+
+}
+
+bool Board::isKing(const Location& loc) const
+{
+    return m_board[loc.x][loc.y]->getCharacter() == Character::KING;
+}
+
+bool Board::isPawn(const Location& loc) const
+{
+    return m_board[loc.x][loc.y]->getCharacter() == Character::PAWN;
+}
+
+bool Board::mate() const
+{
+    auto opponentMoves = getOpponentMoves();
+
+    for (size_t i = 0; i < opponentMoves.size(); ++i)
+    {
+        auto enemyMove = opponentMoves[i];
+        if (m_board[enemyMove.x][enemyMove.y]->isCorrectMove(enemyMove, m_kingLoc[m_playerIsWhite]) &&
+            !isWayBlocked(enemyMove, m_kingLoc[m_playerIsWhite]))
+            return true;
+    }
+
+    return false;
+}
+
+std::vector<Location> Board::getOpponentMoves() const
+{
+    std::vector<Location> opponentMoves;
+
+    for (size_t r = 0; r < ROWS; r++)
+    {
+        for (size_t c = 0; c < COLS; c++)
+        {
+            Location currentLoc = { r, c };
+            if (!isEmpty(currentLoc) && !isCurrentPiecePlayer(currentLoc))
+                opponentMoves.push_back(currentLoc);
+        }
+    }
+    return opponentMoves;
+}
+
+void Board::update(const Location source, const Location destination)
+{
+    if (isKing(source))
+        m_kingLoc[m_board[source.x][source.y]->getColor()] = destination;
+
+    if (isPawn(source))
+    {
+        Pawn* pawn = dynamic_cast<Pawn*>(m_board[source.x][source.y].get());
+        pawn->setFirstMove();
+    }
+        
+    m_board[destination.x][destination.y] = std::move(m_board[source.x][source.y]);

```

Comment:
Instead of using std::move to transfer ownership of the piece from the source position to the destination position, you can use std::exchange to simultaneously move and nullify the source position. you can read about it here: https://en.cppreference.com/w/cpp/utility/exchange